### PR TITLE
Allow permissionless deployments of future pools using minimal proxies

### PR DESCRIPTION
**Pull request by penandlim**
Changes to Swap.sol
* Allow explicit initialization via `initialize` function
* Remove allowlist dependency
* Remove `merkleProof` parameter in addLiquidity function
* Remove `guarded` state variable

Changes to ISwap.sol
* Interface for addLiquidity has changed

Changes to LPToken.sol
* Remove allowlist dependency
* Remove totalminted state variable

### CODE REVIEW

**Comment by penandlim** [comment]
There is a change in the Swap interface 

https://github.com/saddle-finance/saddle-contract/pull/277/files#diff-74f58384743ec3cd5b10e92b80ef0bd78625229463ad739f91901c069c58e6b6L55-L60

addLiquidity function has been modified to exclude the merkleProof parameter.

File: contracts/Swap.sol [hunk 1]
```diff
@@ -124,9 +116,32 @@ contract Swap is OwnerPausable, ReentrancyGuard {
         uint256 _a,
         uint256 _fee,
         uint256 _adminFee,
-        uint256 _withdrawFee,
-        IAllowlist _allowlist
-    ) public OwnerPausable() ReentrancyGuard() {
+        uint256 _withdrawFee
+    ) public {
+        initialize(
```
Comment (prestwich) [inline contracts/Swap.sol:121]: why is this needed? will contracts ever be deployed by constructor?

Comment (penandlim) [reply contracts/Swap.sol:121]: This is for the sake of compatibility. If we ever decide to launch a pool directly from an EOA in 1 transaction, we have the option to do so.

Comment (prestwich) [reply contracts/Swap.sol:121]: hmmm. it's somewhat conceptually messy that the template contract has to be initialized with a full parameter list. Is there a specific use-case for this in mind?

File: test/testUtils.ts [hunk 2]
```diff
@@ -141,6 +141,10 @@ export async function setTimestamp(timestamp: number): Promise<any> {
   return ethers.provider.send("evm_mine", [timestamp])
 }
 
+export async function forceAdvanceOneBlock(): Promise<any> {
+  return ethers.provider.send("evm_mine", [])
+}
+
```
Comment (alphastorm) [inline test/testUtils.ts:147]: should probably refactor all these duplicate calls:

Screen Shot 2021-03-08 at 1 30 55 PM

File: contracts/Swap.sol (hunk 1 above)
Comment (alphastorm) [reply contracts/Swap.sol:121]: let's remove it for now, can always add stuff back later if/when there is an explicit need

File: contracts/SwapDeployer.sol [hunk 3]
```diff
@@ -0,0 +1,43 @@
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "openzeppelin-contracts-3.4/proxy/Clones.sol";
+import "./interfaces/ISwap.sol";
+import "hardhat/console.sol";
```
Comment (alphastorm) [inline contracts/SwapDeployer.sol:6]: should this be removed?

File: test/testUtils.ts [hunk 4]
```diff
@@ -133,21 +133,25 @@ export async function setNextTimestamp(timestamp: number): Promise<any> {
...
+  return forceAdvanceOneBlock(timestamp)
 }
 
-export async function forceAdvanceOneBlock(): Promise<any> {
-  return ethers.provider.send("evm_mine", [])
+export async function forceAdvanceOneBlock(timestamp?: number): Promise<any> {
+  const params = []
+  if (timestamp) {
+    params.push(timestamp)
+  }
```
Comment (alphastorm) [inline test/testUtils.ts:148]: nit:
```js
const params = timestamp ? [timestamp] : []
```

**Review by alphastorm — APPROVED** [review]

**Review by alphastorm — APPROVED** [review]
